timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -837,4 +810,85 @@ void __global__ k_adjust_sample_idxs(
     }
 }
 
+// k_accept_first_valid_move selects the first sample within a batch of samples. Use AtomicMin to
+// find the lowest index sample that was selected, then store that sample in accepted sample field. If no sample is
+// accepted store N in the accepted sample, indicating no valid sample. The accepted sample can also be used as the offset
+// for shifting the noise for backtracking
+template <typename RealType>
+void __global__ k_accept_first_valid_move(
+    const int total_proposals,
+    const int num_target_mols,
+    const int batch_size,
+    const int *__restrict__ noise_offset,        // [1]
+    const int *__restrict__ samples,             // [batch_size]
+    const RealType *__restrict__ before_max,     // [1]
+    const RealType *__restrict__ before_log_sum, // [1]
+    const RealType *__restrict__ after_max,      // [batch_size]
+    const RealType *__restrict__ after_log_sum,  // [batch_size]
+    const RealType *__restrict__ rand,           // [total_proposals]
+    int *__restrict__ accepted_sample            // [1]
+) {
+    __shared__ int selected_idx;
+    assert(blockIdx.x == 0);
```

Comment:
I wonder if we can turn statements like this into compile time `static_asserts`, (blockIdx.x may not be known at compile time, but gridDim.x should be known at compile time?)